src/Processors/Transforms/PartialSortingTransform.cpp

```diff
@@ -66,22 +69,30 @@ void PartialSortingTransform::transform(Chunk & chunk)
             if (less(threshold_block_columns, limit - 1, block_columns, i, description))
             {
                 ++filtered_count;
-                filter[i] = 1;
+                filter[i] = 0;
             }
         }
 
         if (filtered_count)
         {
             for (auto & column : block.getColumns())
             {
-                column = column->filter(filter, filtered_count);
+                column = column->filter(filter, rows_num - filtered_count);
             }
         }
     }
 
     sortBlock(block, description, limit);
 
-    if (limit && limit < block.rows() &&
+    if (!threshold_block_columns.empty())
```

Comment:
Why do we need to extract columns again?
I think it's the same as this code:
https://github.com/ClickHouse/ClickHouse/pull/11291/files#diff-ddd3b9ea171040c44d707fcc8d1ce9fcR61-R64

Reply:
Because `sortBlock()` truncated columns by limit, so we should renew pointers to columns if this isn't the first block.